File: LeetCode/Medium/2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int startValue;
    int destValue;
    
    public String getDirections(TreeNode root, int startValue, int destValue) {
        this.startValue = startValue;
        this.destValue = destValue;
        
        TreeNode lca = lca(root);
        StringBuilder s = new StringBuilder();
        StringBuilder d = new StringBuilder();
        
        search(lca, startValue, s);
        search(lca, destValue, d);

        return "U".repeat(s.length()) + d.toString();
    }
    
    private TreeNode lca(TreeNode node){
        if (node == null) return null;
        
        if (node.val == startValue || node.val == destValue) return node;
        
        TreeNode l = lca(node.left);
        TreeNode r = lca(node.right);
        
        if (l != null && r != null) return node;
        else if (l != null) return l;
        else if (r != null) return r;
        
        return null;
    }
    
    private boolean search(TreeNode node, int v, StringBuilder curr){
        if (node == null) return false;
        
        if (node.val == v) return true;
        
        curr.append("L");
        if (search(node.left, v, curr)) return true;
        
        curr.setLength(curr.length() - 1);
        
        curr.append("R");
        if (search(node.right, v, curr)) return true;
        
        curr.setLength(curr.length() - 1);
        
        return false;
        
    }
}

File: LeetCode/Medium/2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        def lca(node):
            if not node:
                return None
            
            if node.val == startValue or node.val == destValue:
                return node
            
            l = lca(node.left)
            r = lca(node.right)

            if l and r:
                return node
            elif l:
                return l
            else:
                return r

        def search(node, v, curr):
            if not node:
                return False

            if node.val == v:
                return True

            curr.append("L")
            l = search(node.left, v, curr)

            if l:
                return True

            curr.pop()

            curr.append("R")
            r = search(node.right, v, curr)

            if r:
                return True
            
            curr.pop()

            return False

        lca = lca(root)
        s = []
        d = []
        search(lca, startValue, s)
        search(lca, destValue, d)

        return "U" * len(s) + "".join(d)